README.md
# dash-apps
scikit-rf applications written in [plotly dash](https://plotly.com/dash/)

## Time Gate 

![time gate](https://raw.githubusercontent.com/scikit-rf/dash-apps/main/time_gate/time_gate.png)


SNaP/CHANGELOG.md
### V 0.4.0
Move to flask-caching instead of hidden div (2.5x increase in performance)

### V 0.3.0
Initial Time Gating Support in viewer app

### V 0.2.0
Switched to tabular based layout with loaded data always visible,
in preparation for adding new "apps" beyond the snp viewer in the future.

Implemented ability to select specific uploaded networks to plot.


SNaP/tests/test_viewer.py
# Jackson Anderson
import json
import numpy as np
import skrf as rf
from os.path import join, realpath, pardir, abspath


# from hypothesis import given, strategies as st
# from hypothesis.extra import numpy as stnp


# unit_dict = { \
#     'hz': 'Hz', \
#     'khz': 'KHz', \
#     'mhz': 'MHz', \
#     'ghz': 'GHz', \
#     'thz': 'THz' \
#     }
# datatypes = st.one_of(st.integers,st.floats,st.complex_numbers)
# # TODO: fix arguements for st.builds so that it generates correctly
#
# def build_hypothesis_test_network(nports, npoints):
#     return st.builds(
#         rf.Network,
#         name=st.characters,
#         f=stnp.arrays(datatypes, [npoints]),
#         z0=stnp.arrays(datatypes, [st.sampled_from([nports, npoints])]),
#         s=stnp.arrays(datatypes, [npoints, nports, nports]),
#         comments=st.characters,
#         f_unit=st.sampled_from((None, 'hz', 'khz', 'mhz', 'ghz'))
#     )

# test_ntwk = build_hypothesis_test_network(st.integers(min_value=1, max_value=20),
#                                           st.integers(min_value=0, max_value=100000))
#
#
# @given(test_ntwk)
# def test_random_snp(obj):
#     assert False


class TouchstoneEncoder(json.JSONEncoder):
    def default(self, obj):
        if isinstance(obj, np.ndarray):
            return obj.tolist()
        if isinstance(obj, np.complex):
            return np.real(obj), np.imag(obj)  # split into [real, im]
        if isinstance(obj, rf.Frequency):
            return {'flist': obj.f_scaled.tolist(), 'funit': obj.unit}
        return json.JSONEncoder.default(self, obj)


def to_json(network):
    return json.dumps(network.__dict__, cls=TouchstoneEncoder)


def from_json(obj_string):
    obj = json.loads(obj_string)
    ntwk = rf.Network()
    ntwk.variables = obj['variables']
    ntwk.name = obj['name']
    ntwk.comments = obj['comments']
    ntwk.port_names = obj['port_names']
    ntwk.z0 = np.array(obj['_z0'])[..., 0] + np.array(obj['_z0'])[..., 1] * 1j  # recreate complex numbers
    ntwk.s = np.array(obj['_s'])[..., 0] + np.array(obj['_s'])[..., 1] * 1j
    ntwk.frequency = rf.Frequency.from_f(np.array(obj['_frequency']['flist']),
                                         unit=obj['_frequency']['funit'])
    return ntwk


def test_snp_json_roundtrip():
    '''
    Tests if snp object saved to json and reloaded is still the same.
    :return:
    '''
    datafile = r'357_pcb_960978_um_pe_001_25.s3p'
    rootdir = abspath(join(realpath(__file__), pardir, pardir))
    given = rf.Network(join(rootdir, 'example_data', datafile))
    actual = from_json(to_json(given))
    assert actual == given
    assert actual.frequency == given.frequency
    assert actual.name == given.name



SNaP/snap/index.py
# -*- coding: utf-8 -*-
"""
Created on Tue Mar 26 14:57:14 2019

@author: ander906
"""
import sys
from os.path import join, dirname, abspath, basename, realpath
import dash_core_components as dcc
import dash_html_components as html
from dash.dependencies import Input, Output

def isnotebook():
    '''
    Check if in jupyterlab environment.

    From https://stackoverflow.com/questions/15411967/how-can-i-check-if-code-is-executed-in-the-ipython-notebook
    '''
    try:
        shell = get_ipython().__class__.__name__
        if shell == 'ZMQInteractiveShell':
            return True   # Jupyter notebook or qtconsole
        elif shell == 'TerminalInteractiveShell':
            return False  # Terminal running IPython
        else:
            return False  # Other type (?)
    except NameError:
        return False      # Probably standard Python interpreter


in_jupyterlab = isnotebook()
hubzero = False
auth_on = True
flask_debug = False
for arg in sys.argv:  # Check if any hubzero runtime arguments were specified.
    if arg == 'hubzero':
        hubzero = True
    elif arg == 'noauth':
        auth_on = False
    elif arg == "debug":
        flask_debug = True
    else:
        continue
# Handle conditional imports for hubzero hosting
if hubzero:
    from hublib.util import check_access
    from hubzeroapp import app

    if auth_on: check_access(app)
elif in_jupyterlab:
    import jupyterlab_dash
    viewer = jupyterlab_dash.AppViewer()
    from app import app
else:
    from app import app

server = app.server

from apps import app_viewer

if flask_debug:
    from werkzeug.middleware.profiler import ProfilerMiddleware

changes = ''
with open(join(dirname(dirname(abspath(__file__))), 'CHANGELOG.md')) as f:
    changes = changes+f.read()

app.layout = html.Div([
    dcc.Location(id='url', refresh=False),
    html.Div(id='page-content'),
])


@app.callback(Output('page-content', 'children'),
              [Input('url', 'pathname')])
def display_page(pathname):
    try:  # Need to take basename to handle nanohub proxy
        if basename(pathname):
            pathname=basename(pathname)
    except TypeError:  # On start of app, pathname will be NoneType, throwing error unless caught
        pass
    if pathname == 'app_viewer':
        return app_viewer.layout
    else:
        return html.Div([
            html.A(
                children=html.Img(src=app.get_asset_url('github2.svg')),
                href=r'https://github.com/JAnderson419/SNaP',
                style={'fill': r'#151513', 'color': r'#fff', 'position': 'absolute', 'top': 0, 'border': 0, 'right': 0}
            ),
            html.H1('SNaP SnP Utilities'),
            html.H3(children='by Jackson Anderson, ander906@purdue.edu'),
            html.Hr(),
            html.H5('Release Notes'),
            html.Div(
                dcc.Markdown(children=changes),
                style={'border': '2px solid #a3a3c2', 'background-color': r'#f0f0f5',
                       'height': r'10em', 'overflow': 'scroll', 'resize': 'both'}
            ),
            html.Hr(),
            dcc.Link('Go to SnP Viewer.', href='app_viewer'),
            html.A(
                children=html.Img(src=app.get_asset_url('powered_by_scikit-rf.svg')),
                href=r'http://scikit-rf.org',
                style={'position': 'absolute', 'bottom': 0, 'right': 0}
            ),
        ])


def main():
    if hubzero:
        app.run_server(port=8000, host='0.0.0.0')
    elif in_jupyterlab:
        viewer.show(app)
    else:
        if flask_debug:
            print(join(dirname(dirname(realpath(__file__))), 'debug'))
            app.server.config['PROFILE'] = True
            app.server.wsgi_app = ProfilerMiddleware(app.server.wsgi_app,
                                                     restrictions=[30],
                                                     profile_dir=join(dirname(dirname(realpath(__file__))), 'debug'))
            app.run_server(debug=False)
        else:
            app.run_server(debug=True)


if __name__ == '__main__':
    main()


time_gate/time_gate.py
import dash
import dash_core_components as dcc
import dash_html_components as html
from dash.dependencies import Input, Output, State
import plotly.graph_objects as go
import numpy as np


import skrf as rf
from skrf.time import detect_span



external_stylesheets = [
    'https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css',
    'https://codepen.io/chriddyp/pen/bWLwgP.css'
    ]

app = dash.Dash(__name__,external_stylesheets=external_stylesheets )


''' FYI  The device under test was made with this

air = rf.air
air.npoints =1001
dut = air.shunt_delay_load(.2,180)**\
    air.line(100)**\
    air.shunt_delay_load(.4,100)**\
    air.line(4000)**\
    air.shunt(air.load(.7))**\
    air.line(1000)**\
    air.impedance_mismatch(1,3)
    #**air.short()
dut.add_noise_polar(.01,3e-5)
'''
dut = rf.Network('dut.s2p')


# pull out some frequency info to set slider bounds on center/span
freq = dut.frequency
dt = (freq.t_ns[1]-freq.t_ns[0] )
t_max = freq.t_ns.max()


######## ## the APP
app.layout = html.Div(className='page',children=[
    html.Div(className='sub_page',children=[
    #html.Div(className='col-2'),
    html.Div( children=[
    html.H3(className='product',children=[
        'Time Gating',
        html.A(href='http://www.plotly.com',  children=[
        html.Img(src=app.get_asset_url('dash.png'),
                 style={'height':'50px','float':'right'})
        ]),
        html.A(href='http://www.scikit-rf.org',  children=[
            html.Img(src=app.get_asset_url('powered_by_skrf.png'),
                    style={'height':'50px','float':'right'}),
            ]),

    ]),
    html.Div(className='row',children=[
    html.Div(className='col-4',children=[
        html.Span(children=[
        html.Label('S-parameter: ',className='inline'),
        dcc.RadioItems(
            id='sparam-radio',
            options=[
                {'label':'S11','value':'s11'},
                {'label':'S21','value':'s21'},
                {'label':'S12','value':'s12'},
                {'label':'S22','value':'s22'},

                ],
            value='s11',
            labelStyle={'display': 'inline-block', 'margin':'6px'}
            ),
        ]),


        html.Label('Window Type: '),
        dcc.Dropdown(
            id='window-dropdown',
            options=[
                {'label':'kaiser', 'value':'kaiser' },
                {'label':'boxcar', 'value':'boxcar'},
                {'label':'hamming','value':'hamming'}
                ],
            value='hamming',
            ),

        html.Label('Mode:  ',className='inline'),
        dcc.RadioItems(
            className='inline',
            id='mode-radio',
            options=[
                {'label':'pass','value':'bandpass'},
                {'label':'stop','value':'bandstop'},
                ],
            value='bandpass',
            labelStyle={'display': 'inline-block', 'margin':'6px'}
            ),
        ] ),

    html.Div(className='col',children=[
    html.Button(id='autogate-button',children='AutoGate'),
        html.Label('Center Frequency(ns)'),
        dcc.Slider(
            id='center-slider',
            min=-t_max, max=t_max, value=0, step=dt*10,
            marks = {str(int(x)):str(int(x)) for x in np.linspace(-t_max,t_max,13)},
            ),
        html.Label('Span Frequency(ns)'),
        dcc.Slider(
            id='span-slider',
            min=0, max=t_max, value=10, step=dt*10,
            marks = {str(int(x)):str(int(x)) for x in np.linspace(-t_max,t_max,13)},
            ),
        ]),
    ] ),

    html.Div([
        dcc.Graph(id='time-graph',figure={'layout':{'height':300}}),
        dcc.Graph(id='freq-graph',figure={'layout':{'height':400}}),
        ]),
    ]),
    ])])


@app.callback(
    Output('time-graph','figure'),
    Output('freq-graph','figure'),
    Input('sparam-radio', 'value'),
    Input('center-slider', 'value'),
    Input('span-slider', 'value'),
    Input('window-dropdown','value'),
    Input('mode-radio','value'),
    )
def update_figure(sparam, center,span,window,mode):
    '''
    Update figures given a change in s-parameter of interest, or
    gate parameters
    '''

    if window=='kaiser':
        window= ('kaiser',6)
    oneport = dut.__getattribute__(sparam)
    gated = oneport.time_gate(center=center, span=span, window=window, mode=mode)

    # time
    time_fig = go.Figure()
    time_fig.add_trace(go.Line(x=freq.t_ns,
        y=oneport.s_time_db.flatten(), line_color='#204a87',
        name='OG') )

    time_fig.add_trace(go.Line(
        x=freq.t_ns,
        y=gated.s_time_db.flatten(),
        name='gated'))
    dat = oneport.s_time_db.flatten()
    span_height=(dat.min()*1.5, dat.max()*.5)
    time_fig.update_layout(
        shapes=[
        dict(
            type="rect",
            xref="x",
            yref="y",
            x0=center-span/2,
            y0=span_height[0],
            x1=center+span/2,
            y1=span_height[1],
            fillcolor="lightgray",
            opacity=0.8,
            line_width=0,
            layer="below"
        )])


    time_fig.update_layout(
        legend=dict(yanchor="top",y=0.99, xanchor="left",  x=0.01),
        margin=dict(l=20, r=20, t=40, b=20),
        title="Time Domain",
        xaxis_title="Time (ns)",
        yaxis_title="Magnitude (dB)",
        yaxis_range=(-80,0),)

    # frequency
    freq_fig = go.Figure()
    freq_fig.add_trace(go.Line(x=freq.f_scaled,
        y=oneport.s_db.flatten(),
        name='OG',line_color='#204a87'))

    freq_fig.add_trace(go.Line(
        x=freq.f_scaled,
        y=gated.s_db.flatten(),
        name='gated'))


    freq_fig.update_layout(
        legend=dict(yanchor="top",y=0.99, xanchor="left",  x=0.01),
        margin=dict(l=20, r=20, t=40, b=20),
        title="Frequency",
        xaxis_title="Frequency (%s)"%dut.frequency.unit,
        yaxis_title="Magnitude (dB)",
        yaxis_range=(-60,10),
        )
    return time_fig,freq_fig


@app.callback(
    Output('center-slider', 'value'),
    Output('span-slider', 'value'),
    Input('autogate-button', 'n_clicks'),
    Input('sparam-radio', 'value'),
    )
def autogate(autogate,sparam):
    '''
    Autogate chooses center/span values based on basic peak search
    '''
    oneport = dut.__getattribute__(sparam)
    n = oneport.s_time_mag.argmax()
    center = oneport.frequency.t_ns[n]
    span = detect_span(oneport)
    return center,span

if __name__ == '__main__':
    app.run_server(debug=True)


time_gate/ntwk_uploader.py
# -*- coding: utf-8 -*-

# Run this app with `python app.py` and
# visit http://127.0.0.1:8050/ in your web browser.


import base64
import io

import dash
import dash_core_components as dcc
import dash_html_components as html
import plotly.graph_objects as go
from dash.dependencies import Input, Output,State

import plotly.express as px
import pandas as pd
import skrf as rf
#from skrf.instances import air50 as med
import numpy as np

external_stylesheets = [
    'https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css',
    'https://codepen.io/chriddyp/pen/bWLwgP.css'
    ]

app = dash.Dash(__name__,external_stylesheets=external_stylesheets )

app.layout = html.Div([
    dcc.Upload(id='upload-ntwk', children=[
        html.Button('Upload File')
        ]),
    html.Div(id='ntwk-name',children=['Div']),
   ])



@app.callback(
    Output('ntwk-name','children'),
    Input('upload-ntwk', 'contents'),
    State('upload-ntwk', 'filename'),
    State('upload-ntwk', 'last_modified'),
    )
def update_file(contents, filename, dates):
    if contents is not None:
        print(filename)
        content_type, content_string = contents.split(',')
        decoded = base64.b64decode(content_string)
        dut = rf.Network()
        str_io = io.StringIO(decoded.decode('utf-8'))
        str_io.name = filename # due to bug in skrf
        dut.read_touchstone(str_io)
        return str(dut)
    else:
        return 'upload file'



if __name__ == '__main__':
    app.run_server(debug=True)


time_gate/time_gate_uploader.py
import dash
import dash_core_components as dcc
import dash_html_components as html
from dash.dependencies import Input, Output, State
import plotly.graph_objects as go
import numpy as np
import base64
import io

import skrf as rf
from skrf.time import detect_span



external_stylesheets = [
    'https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css',
    'https://codepen.io/chriddyp/pen/bWLwgP.css'
    ]

app = dash.Dash(__name__,external_stylesheets=external_stylesheets )


# FYI  The device under test was made with this
'''
air = rf.air
air.npoints =4001
dut = air.line(1000)**\
    air.shunt_delay_load(.4,100)**\
    air.line(4000)**\
    air.shunt(air.load(.7))**\
    air.line(1000)**\
    air.impedance_mismatch(1,3)
    #**air.short()
dut.add_noise_polar(.01,3e-5)
dut.write_touchstone('dut2.s2p')
'''



# pull out some frequency info to set slider bounds on center/span



######## ## the APP
app.layout = html.Div(className='page',children=[
    html.Div(className='sub_page',children=[
    #html.Div(className='col-2'),
    html.Div( children=[
    html.H3(className='product',children=[
        'Time Gating',
        html.A(href='http://www.plotly.com',  children=[
        html.Img(src=app.get_asset_url('dash3.png'),
                 style={'height':'50px','float':'right'})
        ]),
        html.A(href='http://www.scikit-rf.org',  children=[
            html.Img(src=app.get_asset_url('skrf1.png'),
                    style={'height':'50px','float':'right'}),
            ]),

    ]),
    html.Div(className='row',children=[
    html.Div(className='col-4',children=[
        html.Span(children=[
        html.Label('S-parameter: ',className='inline'),
        dcc.RadioItems(
            id='sparam-radio',
            options=[
                {'label':'S11','value':'s11'},
                {'label':'S21','value':'s21'},
                {'label':'S12','value':'s12'},
                {'label':'S22','value':'s22'},

                ],
            value='s11',
            labelStyle={'display': 'inline-block', 'margin':'6px'}
            ),
        ]),


        html.Label('Window Type: '),
        dcc.Dropdown(
            id='window-dropdown',
            options=[
                {'label':'kaiser', 'value':'kaiser' },
                {'label':'boxcar', 'value':'boxcar'},
                {'label':'hamming','value':'hamming'}
                ],
            value='hamming',
            ),

        html.Label('Mode:  ',className='inline'),
        dcc.RadioItems(
            className='inline',
            id='mode-radio',
            options=[
                {'label':'pass','value':'bandpass'},
                {'label':'stop','value':'bandstop'},
                ],
            value='bandpass',
            labelStyle={'display': 'inline-block', 'margin':'6px'}
            ),
        ] ),

    html.Div(className='col',children=[
    dcc.Upload(id='upload-ntwk', children=[
        html.Button('Upload File')
        ]),
    html.Div(id='ntwk-name',children=['Div']),
    html.Button(id='autogate-button',children='AutoGate'),
        html.Label('Center Frequency(ns)'),
        dcc.Slider(
            id='center-slider',
            #min=-t_max, max=t_max, value=0, step=dt*10,
            #marks = {str(int(x)):str(int(x)) for x in np.linspace(-t_max,t_max,13)},
            ),
        html.Label('Span Frequency(ns)'),
        dcc.Slider(
            id='span-slider',
            #min=0, max=t_max, value=10, step=dt*10,
            #marks = {str(int(x)):str(int(x)) for x in np.linspace(-t_max,t_max,13)},
            ),
        ]),
    ] ),

    html.Div([
        dcc.Graph(id='time-graph',figure={'layout':{'height':300}}),
        dcc.Graph(id='freq-graph',figure={'layout':{'height':400}}),
        ]),
    ]),
    ])])
@app.callback(
    Output('center-slider','min'),
    Output('center-slider','max'),
    Input('upload-ntwk', 'contents'),
    State('upload-ntwk', 'filename'),
    State('upload-ntwk', 'last_modified'),

    )
def update_sliders(contents, filename, dates):
    dut = update_file(contents, filename, dates)
    freq = dut.frequency
    dt = (freq.t_ns[1]-freq.t_ns[0] )
    t_max = freq.t_ns.max()
    return -t_max,t_max

@app.callback(
    Output('time-graph','figure'),
    Output('freq-graph','figure'),
    Input('sparam-radio', 'value'),
    Input('center-slider', 'value'),
    Input('span-slider', 'value'),
    Input('window-dropdown','value'),
    Input('mode-radio','value'),
    Input('upload-ntwk', 'contents'),
    State('upload-ntwk', 'filename'),
    State('upload-ntwk', 'last_modified'),
    )
def update_figure(sparam, center,span,window,mode,contents, filename, dates):
    '''
    Update figures given a change in s-parameter of interest, or
    gate parameters
    '''
    dut = update_file(contents, filename, dates)
    freq = dut.frequency
    dt = (freq.t_ns[1]-freq.t_ns[0] )
    t_max = freq.t_ns.max()



    if window=='kaiser':
        window= ('kaiser',6)
    oneport = dut.__getattribute__(sparam)
    gated = oneport.time_gate(center=center, span=span, window=window, mode=mode)

    # time
    time_fig = go.Figure()
    time_fig.add_trace(go.Line(x=freq.t_ns,
        y=oneport.s_time_db.flatten(), line_color='#204a87',
        name='OG') )

    time_fig.add_trace(go.Line(
        x=freq.t_ns,
        y=gated.s_time_db.flatten(),
        name='gated'))
    dat = oneport.s_time_db.flatten()
    span_height=(dat.min()*1.5, dat.max()*.5)
    time_fig.update_layout(
        shapes=[
        dict(
            type="rect",
            xref="x",
            yref="y",
            x0=center-span/2,
            y0=span_height[0],
            x1=center+span/2,
            y1=span_height[1],
            fillcolor="lightgray",
            opacity=0.8,
            line_width=0,
            layer="below"
        )])


    time_fig.update_layout(
        legend=dict(yanchor="top",y=0.99, xanchor="left",  x=0.01),
        margin=dict(l=20, r=20, t=40, b=20),
        title="Time Domain",
        xaxis_title="Time (ns)",
        yaxis_title="Magnitude (dB)",
        yaxis_range=(-80,0),)

    # frequency
    freq_fig = go.Figure()
    freq_fig.add_trace(go.Line(x=freq.f_scaled,
        y=oneport.s_db.flatten(),
        name='OG',line_color='#204a87'))

    freq_fig.add_trace(go.Line(
        x=freq.f_scaled,
        y=gated.s_db.flatten(),
        name='gated'))


    freq_fig.update_layout(
        legend=dict(yanchor="top",y=0.99, xanchor="left",  x=0.01),
        margin=dict(l=20, r=20, t=40, b=20),
        title="Frequency",
        xaxis_title="Frequency (%s)"%dut.frequency.unit,
        yaxis_title="Magnitude (dB)",
        yaxis_range=(-60,10),
        )
    return time_fig,freq_fig


@app.callback(
    Output('center-slider', 'value'),
    Output('span-slider', 'value'),
    Input('autogate-button', 'n_clicks'),
    Input('sparam-radio', 'value'),
    )
def autogate(autogate,sparam):
    '''
    Autogate chooses center/span values based on basic peak search
    '''
    oneport = dut.__getattribute__(sparam)
    n = oneport.s_time_mag.argmax()
    center = oneport.frequency.t_ns[n]
    span = detect_span(oneport)
    return center,span



def update_file(contents, filename, dates):
    if contents is not None:
        print(filename)
        content_type, content_string = contents.split(',')
        decoded = base64.b64decode(content_string)
        dut = rf.Network()
        str_io = io.StringIO(decoded.decode('utf-8'))
        str_io.name = filename # due to bug in skrf
        dut.read_touchstone(str_io)
        return dut

    else:
        return rf.Network('dut.s2p')



if __name__ == '__main__':
    app.run_server(debug=True)
